File: Striver/DP/Q47.py
# count number of longest increasing subsequence

def LIS_count(arr,n):
    dp = [1 for _ in range(n)]
    cnt = [1 for _ in range(n)]
    for ind in range(1,n):
        for prev_ind in range(ind):
            if arr[ind] > arr[prev_ind] and dp[ind] < dp[prev_ind] + 1:
                dp[ind] = dp[prev_ind] + 1
                cnt[ind] = cnt[prev_ind]
            elif arr[ind] > arr[prev_ind] and dp[ind] == dp[prev_ind] + 1:
                cnt[ind] += cnt[prev_ind]
        
    mx = max(dp)
    count = 0
    for i in range(n):
        if dp[i] == mx:
            count += cnt[i]
    return count

arr = [1, 3, 5, 4, 7]
n = len(arr)
print(LIS_count(arr, n))

File: Striver/DP/Q47.java
package Striver.DP;

import java.util.Arrays;

public class Q47{
    // count number of longest increasing subsequence
    public int LISCount(int[] arr, int n){
        int[] dp = new int[n];
        int[] cnt = new int[n];
        Arrays.fill(dp, 1);
        Arrays.fill(cnt, 1);
        int max = 1;
        for(int i = 1; i < n; i++){
            for(int j = 0; j < i; j++){
                if(arr[i] > arr[j] && dp[i] < dp[j]+1){
                    dp[i] = dp[j]+1;
                    cnt[i] = cnt[j];
                }
                else if(arr[i] > arr[j] && dp[i] == dp[j]+1){
                    cnt[i] += cnt[j];
                }    
                
            }
            max = Math.max(max, dp[i]);
        }
        int ans = 0;
        for(int i = 0; i < n; i++){
            if(dp[i] == max) ans += cnt[i];
        }
        return ans;
    }
}